Plugins/GeoModel/src/GeoModelDetectorSurfaceFactory.cpp

```diff
@@ -81,62 +97,100 @@ void Acts::GeoModelDetectorSurfaceFactory::construct(
-          }
-          break;
+          convertSensitive(subvol.volume, transform, cache.sensitiveSurfaces);
         }
-      }
 
-      if (!success) {
-        ACTS_DEBUG(name << " (" << vname << " / " << recType(shape)
-                        << ") could not be converted by any converter");
       } else {
-        volNameStems.emplace(name.substr(0, 6));
-        materials.emplace(fpv->getLogVol()->getMaterial()->getName());
+        // convert the full phys vol
+        convertSensitive(physVol, fpv->getAbsoluteTransform(nullptr),
+                         cache.sensitiveSurfaces);
       }
     }
+
     ACTS_VERBOSE("Found " << qFPV.size()
                           << " full physical volumes matching the query.");
-
-    auto streamVec = [](const auto &v) {
-      std::stringstream ss;
-      for (const auto &el : v) {
-        ss << el << " ";
-      }
-      return ss.str();
-    };
-
-    ACTS_INFO("Converted volumes (stems): " << streamVec(volNameStems));
-    ACTS_INFO("Materials of converted volumes: " << streamVec(materials));
   }
+
   ACTS_DEBUG("Constructed "
              << cache.sensitiveSurfaces.size() << " sensitive elements and "
              << cache.passiveSurfaces.size() << " passive elements");
 }
+
+void Acts::GeoModelDetectorSurfaceFactory::convertSensitive(
+    PVConstLink geoPV, const Acts::Transform3 &transform,
+    std::vector<Acts::GeoModelSensitiveSurface> &sensitives) {
+  const GeoLogVol *logVol = geoPV->getLogVol();
+  const GeoShape *shape = logVol->getShape();
+  int shapeId = shape->typeID();
+  std::string name = logVol->getName();
+  std::shared_ptr<const Acts::IGeoShapeConverter> converter =
+      Acts::GeoShapesConverters(shapeId);
+  if (converter != nullptr) {
+    auto converted = converter->toSensitiveSurface(geoPV, transform);
+    if (converted.ok()) {
+      sensitives.push_back(converted.value());
+      const auto &[el, sf] = converted.value();
+
+      ACTS_VERBOSE("(successfully converted: "
+                   << name << " / " << recType(*shape) << " / "
+                   << logVol->getMaterial()->getName() << ")");
+
+      if (!(el && sf)) {
+        throw std::runtime_error("something is nullptr");
+      }
+      return;
+    }
+    ACTS_DEBUG(name << " / " << recType(*shape)
+                    << ") could not be converted by any converter");
+  }
+  ACTS_DEBUG("converter is a nullptr");
+  return;
+}
+
+std::vector<GeoChildNodeWithTrf>
+Acts::GeoModelDetectorSurfaceFactory::findAllSubVolumes(
+    PVConstLink geoPV,
+    std::function<bool(std::string, PVConstLink)> matchFunc) {
+  const std::vector<GeoChildNodeWithTrf> children =
+      getChildrenWithRef(geoPV, false);
+  std::vector<GeoChildNodeWithTrf> foundVols{};
+
+  for (const auto &child : children) {
+    if (matchFunc(child.nodeName, child.volume)) {
+      ACTS_VERBOSE("The subvol" << child.nodeName << "mathes the queries");
+      foundVols.push_back(child);
+    }
+    if (!child.volume->getNChildVols()) {
```

Comment:
Probably also clang-tidy will complain about this, but we should not implicit convert int to bool like this.
```suggestion
    if (child.volume->getNChildVols() != 0) {
```